File: Java/src/Color/ColorRGB.java
package Color;

import Interface.ICalculable;

public class ColorRGB implements ICalculable<ColorRGB> {
    private short red;
    private short green;
    private short blue;

    public ColorRGB() {
        red = 0;
        green = 0;
        blue = 0;
    }

    public ColorRGB(short r) {
        red = (short) (Math.abs(r) % 256);
        green = 0;
        blue = 0;
    }

    public ColorRGB(short r, short g) {
        this(r);
        green = (short) (Math.abs(g) % 256);
        blue = 0;
    }

    public ColorRGB(short r, short g, short b) {
        this(r, g);
        blue = (short) (Math.abs(b) % 256);
    }

    public ColorRGB(ColorRGB crgb) {
        red = crgb.red;
        green = crgb.green;
        blue = crgb.blue;
    }

    public short getRed() {
        return red;
    }

    public short getGreen() {
        return green;
    }

    public short getBlue() {
        return blue;
    }

    public ColorRGB sum(ColorRGB crgb) {
        return new ColorRGB((short) (red + crgb.red), (short) (green + crgb.green), (short) (blue + crgb.blue));
    }

    public ColorRGB sub(ColorRGB crgb) {
        return new ColorRGB((short) (red - crgb.red), (short) (green - crgb.green), (short) (blue - crgb.blue));
    }

    public ColorRGB mul(ColorRGB crgb) {
        return new ColorRGB((short) (red * crgb.red), (short) (green * crgb.green), (short) (blue * crgb.blue));
    }

    public ColorRGB div(ColorRGB crgb) {
        if (crgb.red == 0 || crgb.green == 0 || crgb.blue == 0) {
            return new ColorRGB(this);
        }

        return new ColorRGB((short) (red / crgb.red), (short) (green / crgb.green), (short) (blue / crgb.blue));
    }

    public boolean equals(Object obj) {
        if (this == obj)
            return true;
        if (obj != null && obj.getClass() == getClass()) {
            return red == ((ColorRGB) obj).red && green == ((ColorRGB) obj).green
                    && blue == ((ColorRGB) obj).blue;
        }
        return false;
    }

    public String toString() {
        return "ColorRGB(" + red + "," + green + "," + blue + ")";
    }
}

File: cpp/KalkCpp/Color/colorrgb.cpp
#include "colorrgb.h"
#include "math.h"
#include <typeinfo>

ColorRGB::ColorRGB(unsigned short r, unsigned short g, unsigned short b) : red(r), green(g), blue(b)
{
    red = abs(red) % 256;
    green = abs(green) % 256;
    blue = abs(blue) % 256;
}


unsigned short ColorRGB::getRed() const
{
    return red;
}



unsigned short ColorRGB::getGreen() const
{
    return green;
}



ColorRGB* ColorRGB::clone() const
{
    return new ColorRGB(*this);
}



unsigned short ColorRGB::getBlue() const
{
    return blue;
}



bool ColorRGB::operator==(const ColorRGB& crgb) const
{
    if(this == &crgb)
        return true;
    if(&crgb != 0 && typeid(*this) == typeid(crgb))
    {
        return red == crgb.red && green == crgb.green && blue == crgb.blue;
    }
    return false;
}








ColorRGB* ColorRGB::operator-(const ColorRGB& crgb) const
{
    return new ColorRGB(red-crgb.red, green-crgb.green, blue-crgb.blue);
}



ColorRGB* ColorRGB::operator+(const ColorRGB& crgb) const
{
    return new ColorRGB(red+crgb.red, green+crgb.green, blue+crgb.blue);
}



ColorRGB* ColorRGB::operator*(const ColorRGB& crgb) const
{
    return new ColorRGB(red*crgb.red, green*crgb.green, blue*crgb.blue);
}



ColorRGB* ColorRGB::operator/(const ColorRGB& crgb) const
{
     if(crgb.red == 0 || crgb.green == 0 || crgb.blue == 0){
         return this->clone();
     }
     return new ColorRGB(red/crgb.red, green/crgb.green, blue/crgb.blue);
}



void ColorRGB::printColor() const
{
    std::cout << "ColorRGB(" + std::to_string(red) + "," + std::to_string(green) + "," + std::to_string(blue) + ")";
}



std::ostream& operator<<(std::ostream& os, const ColorRGB& crgb)
{
    return os<<"("<<crgb.red<<","<<crgb.green<<","<<crgb.blue<<")";
}



ColorRGB::~ColorRGB()
{

}
